### ISSUE TITLE
PLL : keep track of dphi

### ISSUE THREAD

**Issue by pfeatherstone** [jgaeddert/liquid-dsp#257]
So I've played with the different ways of doing PLL in liquid.
There are two methods I can see:
- use NCO object which uses a very simple model
- use method described in https://liquidsdr.org/doc/tutorial-pll/ which uses an active lag design.

Now the first method is great since it's quite simple and it estimates both ```phi``` and ```dphi```. So once you've locked onto a carrier, you can adjust frequency as well as phase.

The second seems to be way more accurate but can only track phase. So you can't really lock onto a signal, since when you no longer have any "truth" data, you won't be able to adjust frequency.

Is there a way to enhance the active lag design to also keep track of frequency/dphi?

Cheers

**Comment by pfeatherstone** [comment]
Do you have to combine both into a single loop?
Maybe use the Costas loop as your main algorithm but use the other algorithm to refine phase when you have truth data?

**Comment by pfeatherstone** [comment]
I'm not an electrical engineer so what I'm about to suggest could be naive. It seems that all the PLL is doing is smoothing phase error over time using a filter, in both cases above, using an IIR filter. Why can't we do the same with an averaging filter to estimate phase error and difference filter to estimate frequency error? How about that?

**Comment by pfeatherstone** [comment]
@jgaeddert any thoughts ?

**Comment by jgaeddert** [comment]
It's not a naive thought that the PLL is smoothing the phase error over time. The important piece is balancing the filter gain to keep the loop in its stable region. You can easily extend this to higher order by incorporating frequency offset estimates. The `nco` object in _liquid_ actually already does this.

**Comment by jgaeddert** [comment]
See the example here: https://liquidsdr.org/blog/pll-simple-howto/

**Comment by pfeatherstone** [comment]
Yeah so that's what I was using but it takes a lot of iterations for it to converge well. Will any kind of smoothing do ?

**Comment by jgaeddert** [comment]
I think you'll find you can't set the values arbitrarily otherwise the loop won't be stable; you need to balance between the phase and frequency gains. You can increase the loop bandwidth to reduce convergence time. Honestly a Kalman filter is probably near-optimal, but without implementing it I'm guessing it would have similar performance.

Feel free to play around with the numbers and see if you can find something that works better.

**Comment by pfeatherstone** [comment]
Are Kalman filters used in PLLs ? I love the sound of that

**Comment by jgaeddert** [comment]
They certainly can be, and I'm sure it's been done before. It's non-linear, of course, so you'd have to use an extended Kalman filter at the least.

**Comment by pfeatherstone** [comment]
There are loads of papers on this but no open source implementation as far as i can see.

**Comment by pfeatherstone** [comment]
This would be an awesome feature in liquid, even if it was for demo purposes

**Comment by pfeatherstone** [comment]
https://hal.archives-ouvertes.fr/hal-03203769/document

**Comment by jgaeddert** [comment]
Great reference! And should work very well for non-linear frequency offsets as well as timing loop control. I've worked on extended Kalman filters before, and have been interested in applying it to tracking loops.

Also, this article is one of the rare exceptions to [Bettridge's Law](https://en.wikipedia.org/wiki/Betteridge's_law_of_headlines)

**Comment by noonafter** [comment]
In NCO object, NCO(_pll_step) method adjust frequency first, then adjust phase. However, **from the perspective of sequential circuits**, the frequency and phase should be adjust synchronously, which means when phase is adjust, the frequency remains the same as the previous value. So mehtod NCO(_adjust_phase) should be called first, then NCO(_step), and finally NCO(_adjust_frequency). If we do this, NCO(_step) won't be called again, just like the example in  #https://liquidsdr.org/doc/nco/#section-nco-pll-interface.

**Comment by noonafter** [comment]
In NCO object, NCO(_pll_step) method adjust frequency first, then adjust phase. However, from the perspective of sequential circuits, the frequency and phase should be adjust simultaneously, which means when phase is adjust, the frequency remains the same as the previous value. So mehtod NCO(_adjust_phase) should be called first, then NCO(_step), and finally NCO(_adjust_frequency). If we do this, NCO(_step) won't be called again, as the example in https://liquidsdr.org/doc/nco/#section-nco-pll-interface.
